include/gauge_path_helper.cuh

```diff
@@ -0,0 +1,115 @@
+#pragma once
+
+#include <gauge_field_order.h>
+#include <quda_matrix.h>
+#include <index_helper.cuh>
+#include <kernel.h>
+#include <thread_array.h>
+
+namespace quda {
+
+  template <int dim_>
+  struct paths {
+    static constexpr int dim = dim_;
+    const int num_paths;
+    const int max_length;
+    int *input_path[dim];
+    const int *length;
+    const double *path_coeff;
+    int *buffer;
+    int count;
+
+    paths(int ***input_path, int *length_h, double *path_coeff_h, int num_paths, int max_length) :
+      num_paths(num_paths),
+      max_length(max_length),
+      count(0)
+    {
+      // create path struct in a single allocation
+      size_t bytes = dim * num_paths * max_length * sizeof(int) + num_paths * sizeof(int);
+      int pad = ((sizeof(double) - bytes % sizeof(double)) % sizeof(double))/sizeof(int);
+      bytes += pad*sizeof(int) + num_paths*sizeof(double);
+
+      buffer = static_cast<int*>(pool_device_malloc(bytes));
+      int *path_h = static_cast<int*>(safe_malloc(bytes));
+      memset(path_h, 0, bytes);
+
+      for (int dir=0; dir<dim; dir++) {
+        // flatten the input_path array for copying to the device
+        for (int i = 0; i < num_paths; i++) {
+          for (int j = 0; j < length_h[i]; j++) {
+            path_h[dir * num_paths * max_length + i * max_length + j] = input_path[dir][i][j];
+            if (dir==0) count++;
+          }
+        }
+      }
+
+      // length array
+      memcpy(path_h + dim * num_paths * max_length, length_h, num_paths*sizeof(int));
+
+      // path_coeff array
+      memcpy(path_h + dim * num_paths * max_length + num_paths + pad, path_coeff_h, num_paths*sizeof(double));
+
+      qudaMemcpy(buffer, path_h, bytes, qudaMemcpyHostToDevice);
+      host_free(path_h);
+
+      // finally set the pointers to the correct offsets in the buffer
+      for (int d=0; d < dim; d++) this->input_path[d] = buffer + d*num_paths*max_length;
+      length = buffer + dim*num_paths*max_length;
+      path_coeff = reinterpret_cast<double*>(buffer + dim * num_paths * max_length + num_paths + pad);
+    }
+
+    void free() {
```

Comment:
Is there some reason why we can't put the `free` function in the destructor to avoid having to call this?

Reply:
(I know this is legacy behaviour, but I can't remember)

Reply:
The theoretical issue with putting it in a destructor is if this got moved around as a copy by value as opposed to by reference. Since this object gets passed directly as a kernel argument, we can't just wrap the pointers in smart pointers to avoid this issue. I guess we could avoid it by creating a smart pointer (with a custom deallocator) outside the function, with some `static` function that'll do the math for you, and then passing the raw pointer in, but it's just extra plumbing. Happy to do it if you think that's best.

Reply:
Thanks for the reminder on this, the theoretical issue is exactly the issue at hand.  Just leave it as is for now (resolving).

Reply:
as I just had the same thought: This calls for being able to use some smart ptr ... Mayeb for now add a comment?